contracts/GoodGhostingPolygonCurve.sol

```diff
@@ -0,0 +1,674 @@
+    //     @param _segmentLength Lenght of each segment, in seconds (i.e., 180 (sec) => 3 minutes).
+    //     @param _segmentPayment Amount of tokens each player needs to contribute per segment (i.e. 10*10**18 equals to 10 DAI - note that DAI uses 18 decimal places).
+    //     @param _earlyWithdrawalFee Fee paid by users on early withdrawals (before the game completes). Used as an integer percentage (i.e., 10 represents 10%).
+    //     @param _customFee performance fee charged by admin. Used as an integer percentage (i.e., 10 represents 10%). Does not accept "decimal" fees like "0.5".
+    //     @param _dataProvider id for getting the data provider contract address 0x1 to be passed.
+    //     @param _maxPlayersCount max quantity of players allowed to join the game
+    //     @param _incentiveToken optional token address used to provide additional incentives to users. Accepts "0x0" adresses when no incentive token exists.
+    //  */
+    constructor(
+        IERC20 _inboundCurrency,
+        ICurvePool _pool,
+        int128 _inboundTokenIndexInt,
+        uint256 _inboundTokenIndexUint,
+        uint256 _poolType,
+        ICurveGauge _gauge,
+        uint256 _segmentCount,
+        uint256 _segmentLength,
+        uint256 _segmentPayment,
+        uint256 _earlyWithdrawalFee,
+        uint256 _customFee,
+        uint256 _maxPlayersCount,
+        IERC20 _curve,
+        IERC20 _matic,
+        IERC20 _incentiveToken
+    ) public {
+        require(
+            _customFee <= 20,
+            "_customFee must be less than or equal to 20%"
+        );
+        require(
+            _earlyWithdrawalFee <= 10,
+            "_earlyWithdrawalFee must be less than or equal to 10%"
+        );
+        require(
+            _earlyWithdrawalFee > 0,
+            "_earlyWithdrawalFee must be greater than zero"
+        );
+        require(
+            _maxPlayersCount > 0,
+            "_maxPlayersCount must be greater than zero"
+        );
+        require(
+            address(_inboundCurrency) != address(0),
+            "invalid _inboundCurrency address"
+        );
+        require(_segmentCount > 0, "_segmentCount must be greater than zero");
+        require(_segmentLength > 0, "_segmentLength must be greater than zero");
+        require(
+            _segmentPayment > 0,
+            "_segmentPayment must be greater than zero"
+        );
+        require(address(_pool) != address(0), "invalid _pool address");
+        require(address(_gauge) != address(0), "invalid _gauge address");
+        require(address(_curve) != address(0), "invalid _curve address");
+        require(address(_matic) != address(0), "invalid _matic address");
+        // Initializes default variables
+        pool = _pool;
+        gauge = _gauge;
+        curve = _curve;
+        matic = _matic;
+        inboundTokenIndexInt = _inboundTokenIndexInt;
+        inboundTokenIndexUint = _inboundTokenIndexUint;
```

Comment:
can't we really cast from uint to int (or vice versa)? I don't think tokenIndex needs to stored as a uint256 (if it only stores the token index), and then we can probably cast it to `int128` or `uint256` at the time we pass it to curve pool, can't we?